src/data-structures/hash-table.js

```diff
@@ -0,0 +1,213 @@
+/**
+ * Hash Table
+ *
+ * An associative array, that can map keys
+ * (strings and numbers) to values in O(1).
+ *
+ * @example
+ * var hash = require('path-to-algorithms/src/data-structures'+
+ * '/hash-table');
+ * var hashTable = new hash.Hashtable();
+ *
+ * hashTable.put(10, 'value');
+ * hashTable.put('key', 10);
+ *
+ * console.log(hashTable.get(10)); // 'value'
+ * console.log(hashTable.get('key')); // 10
+ *
+ * hashTable.remove(10);
+ * hashTable.remove('key');
+ *
+ * console.log(hashTable.get(10)); // undefined
+ * console.log(hashTable.get('key')); // undefined
+ *
+ * @module data-structures/hash-table
+*/
+(function (exports) {
+  'use strict';
+
+  exports.Node = function (key, data) {
+    this.key = key;
+    this.data = data;
+    this.next = undefined;
+    this.prev = undefined;
+  };
+
+  exports.Hashtable = function () {
+    this.buckets = [];
+    // The higher the bucket count; less likely for collisions.
+    this.maxBucketCount = 100;
+  };
+
+  /*
+    Using simple non-crypto x->integer based hash.
+  */
+  exports.Hashtable.prototype.hashCode = function (val) {
+    var i;
+    var hashCode = 0;
+    var character;
+
+    // If value to be hashed is already an integer, return it.
+    if (val.length === 0 || val.length === undefined) {
```

Comment:
I'd use `val !== undefined && val !== null && typeof val.valueOf() === 'number'` here, otherwise I'd cast `val` to string using `'' + val`, which should work in case `val` is `undefined` or `null`.